hail/python/hail/expr/expressions/typed_expressions.py

```diff
@@ -815,6 +815,31 @@ def __getitem__(self, item):
         idxs = to_expr(item, ir.tarray(ir.tint64))
         return construct_expr(ir.NDArrayRef(self._ir, idxs._ir), self._type.element_type)
 
+    @typecheck_method(f=func_spec(1, expr_any))
+    def map(self, f):
+        """Transform each element of an NDArray.
+
+        Parameters
+        ----------
+        f : function ( (arg) -> :class:`.Expression`)
+            Function to transform each element of the NDArray.
+
+        Returns
+        -------
+        :class:`.NDArrayExpression`.
+            NDArray where each element has been transformed according to `f`.
+        """
+
+        element_type, ndim = self._type.element_type, self._type.ndim
+        ndarray_map = self._ir_lambda_method(NDArrayMap, f, element_type, lambda t: self._type.__class__(t, ndim))
```

Comment:
Wow so right that's much better.

Reply:
I believe this was done at one point because it makes it possible for someone to subclass our types/irs/classes, but I don't have any compulsion to support that use case.